src/hotspot/share/gc/g1/g1ServiceThread.cpp

```diff
@@ -24,7 +24,11 @@
 
 #include "precompiled.hpp"
 #include "gc/g1/g1ServiceThread.hpp"
+#include "gc/g1/heapRegion.inline.hpp"
+#include "gc/g1/heapRegionRemSet.inline.hpp"
+#include "gc/shared/suspendibleThreadSet.hpp"
```

Comment:
I think all changes to g1ServiceThread.* could be reverted now when you added the state to the new task instead. Adding `is_enqueued()` to not check `next() == NULL` is a good change, but it should be made on its own.

Reply:
I agree. Note that I think that these changes fix real issues (no virtual destructor with virtual methods) too.